### PR TITLE
Refactoring 🍩

### PR DESCRIPTION
**Pull request by laurmaedje**
I _might_ do some more work on `biblatex`, but I figured this should be worthwile to merge as is.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -209,136 +217,287 @@ impl IntoIterator for Bibliography {
...
+        if let Some(keys) = self.get_as::<Vec<String>>("xdata") {
+            for key in keys {
+                refs.extend(bib.get(&key).cloned());
+            }
         }
 
-        res
+        for mut crossref in refs {
+            crossref.resolve_crossrefs(bib);
+            self.resolve_single_crossref(crossref);
```
Comment (reknih) [inline src/lib.rs:478]: We still do not check for circular references 😭

File: src/lib.rs [hunk 2]
```diff
@@ -423,368 +582,163 @@ impl Entry {
...
+                self.get($field)$(?.parse::<$ret>())?
             }
-        )*
+
+            fields!(@set $name => $field, $($ret)?);
+        })*
     };
 
-    (single_set $name:ident => $field_name:expr, ) => {
+    (@ret) => {&[Chunk]};
```
Comment (reknih) [inline src/lib.rs:605]: Ret stands for Return?

File: src/mechanics.rs [hunk 3]
```diff
@@ -55,24 +55,18 @@ pub enum EntryType {
 /// Describes the optionality mode of the `author` and `editor` fields.
 #[derive(Clone, Debug)]
 pub enum AuthorMode {
+    /// Neither of the fields are required to be set.
+    NoneRequired,
+    /// At least one of the fields must be present.
+    OneRequired,
```
Comment (reknih) [inline src/mechanics.rs:61]: That's way clearer than `AuthorEditorRequired`.

**Review by reknih — APPROVED** [review]
Looks good, thanks! What kind of work did you plan on maybe doing?

File: bench/src/bench.rs [hunk 4]
```diff
@@ -11,12 +12,16 @@ fn benchmarks(c: &mut Criterion) {
...
 
-    bench!("parse-gral": Bibliography::from_str(GRAL, true));
+    bench!("parse-gral": Bibliography::parse(GRAL));
     bench!("get-last-gral": bib.get("adedeji2017"));
-    bench!("get-author-gral": entry.get_author());
+    bench!("get-author-gral": entry.author());
+
+    let bib = Bibliography::parse(CROSS);
+    let entry = bib.get("issue201").unwrap();
+    bench!("get-date-cross": entry.date());
```
[the remark this answers is not in the document]
Comment (laurmaedje) [reply bench/src/bench.rs:24]: Yes.

File: src/lib.rs [hunk 5]
```diff
@@ -1,206 +1,214 @@
...
-                cite_key: entry.cite_key.to_string(),
-                entry_type: EntryType::robust_from_str(entry.entry_type),
+            res.insert(Entry {
+                key: entry.key.to_string(),
+                entry_type: EntryType::new(entry.entry_type),
                 fields: entry
                     .fields
                     .into_iter()
-                    .map(|(key, value)| (key.to_string(), resolve(value, abbreviations)))
+                    .map(|(key, value)| (key.to_string(), resolve::resolve(value, abbr)))
```
[the remark this answers is not in the document]
Comment (laurmaedje) [reply src/lib.rs:88]: Well, that's certainly preference, but as a rule of thumb I mostly import types and traits, but use functions from their module (e.g. `use std::cmp` + `cmp::max`, but `use std::cmp::Ordering`). I think that is a rather common practice for Rust, though I'm not totally sure.

File: src/lib.rs (hunk 2 above)
Comment (laurmaedje) [reply src/lib.rs:605]: Yes.

File: src/mechanics.rs (hunk 3 above)
Comment (laurmaedje) [reply src/mechanics.rs:61]: I'm still not super happy with the enum, but I didn't find a better solution and it's not part of the public API anymore.

**Comment by laurmaedje** [comment]
> What kind of work did you plan on maybe doing?

The circular reference thing, maybe refactoring the low-level parser a bit and adding more tests and possibly looking whether we really need chrono & regex.

**Pull request merged**